### ISSUE TITLE
Improve testing infra for fetchTree/fetchGit

### ISSUE THREAD

**Issue by DavHau** [NixOS/nix#9388]
**Is your feature request related to a problem? Please describe.**
None of the code paths that could trigger actual networking seem to be tested as of now, limiting the test coverage to only `file://` based actions.
I would like to add tests for potential bugs like [this](https://github.com/NixOS/nix/pull/9240#discussion_r1398245212).

**Describe the solution you'd like**
Git already offers the [git-http-backend](https://git-scm.com/docs/git-http-backend) CGI script. Therefore, all that's needed is a cgi capable http server. Python could be used, as it has all of that already built in, but having python as a dependency might be undesired due to bootstrapping complexity. Maybe there is a simpler cpp based http server that can be used.

**Describe alternatives you've considered**
Not testing network related code paths -> Instead rely on humans to find bugs during code review -> slow and unreliable.

**Additional context**

**Priorities**

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

**Comment by roberth** [comment]
I'd be happy to upstream into nixpkgs these modules that help set up a gitea. https://github.com/hercules-ci/hercules-ci-effects/tree/master/effects/testsupport

gitea is fairly quick to launch (unlike, say, gitlab) VM+gitea take [about 10 s](https://hercules-ci.com/accounts/github/hercules-ci/derivations/%2Fnix%2Fstore%2Fbvv9ppnkmh3rl8lbyj4s87hqyh96s3xh-vm-test-run-git-update.drv/log?via-job=1aa0286c-3ccc-4f92-93a1-30e814964a76).

OTOH these a lot of value in keeping the test setup minimal.

> might be undesired due to bootstrapping complexity.

Bootstrapping should be done without tests.
1. bootstrap without tests
2. have a Nix
3. bootstrap with tests, using Nix to get the test deps

Also this could be a `checks` test instead of an `installCheckPhase` test.

**Comment by DavHau** [comment]
Yes, it would be nice if you could upstream these modules. I played around a bit wit the python http server, but it seems like advanced path rewriting capabilities are required in order to use git-http-server. Just using gitea might be simpler.
